tpmutil/encoding.go

```diff
@@ -23,93 +23,28 @@ import (
 	"reflect"
 )
 
-// lengthPrefixSize is the size in bytes of length prefix for byte slices.
-//
-// In TPM 1.2 this is 4 bytes.
-// In TPM 2.0 this is 2 bytes.
-var lengthPrefixSize int
-
-const (
-	tpm12PrefixSize = 4
-	tpm20PrefixSize = 2
+var (
+	selfMarshalerType = reflect.TypeOf((*SelfMarshaler)(nil)).Elem()
+	handlesAreaType   = reflect.TypeOf((*[]Handle)(nil))
 )
 
-// UseTPM12LengthPrefixSize makes Pack/Unpack use TPM 1.2 encoding for byte
-// arrays.
-func UseTPM12LengthPrefixSize() {
-	lengthPrefixSize = tpm12PrefixSize
-}
-
-// UseTPM20LengthPrefixSize makes Pack/Unpack use TPM 2.0 encoding for byte
-// arrays.
-func UseTPM20LengthPrefixSize() {
-	lengthPrefixSize = tpm20PrefixSize
-}
-
-// packedSize computes the size of a sequence of types that can be passed to
-// binary.Read or binary.Write.
-func packedSize(elts ...interface{}) (int, error) {
-	var size int
-	for _, e := range elts {
-		marshaler, ok := e.(SelfMarshaler)
-		if ok {
-			size += marshaler.TPMPackedSize()
-			continue
-		}
-		v := reflect.ValueOf(e)
-		switch v.Kind() {
-		case reflect.Ptr:
-			s, err := packedSize(reflect.Indirect(v).Interface())
-			if err != nil {
-				return 0, err
-			}
-
-			size += s
-		case reflect.Struct:
-			for i := 0; i < v.NumField(); i++ {
-				s, err := packedSize(v.Field(i).Interface())
-				if err != nil {
-					return 0, err
-				}
-
-				size += s
-			}
-		case reflect.Slice:
-			switch s := e.(type) {
-			case []byte:
-				size += lengthPrefixSize + len(s)
-			case RawBytes:
-				size += len(s)
-			default:
-				return 0, fmt.Errorf("encoding of %T is not supported, only []byte and RawBytes slices are", e)
-			}
-		default:
-			s := binary.Size(e)
-			if s < 0 {
-				return 0, fmt.Errorf("can't calculate size of type %T", e)
-			}
-
-			size += s
-		}
-	}
-
-	return size, nil
-}
-
 // packWithHeader takes a header and a sequence of elements that are either of
 // fixed length or slices of fixed-length types and packs them into a single
 // byte array using binary.Write. It updates the CommandHeader to have the right
 // length.
 func packWithHeader(ch commandHeader, cmd ...interface{}) ([]byte, error) {
 	hdrSize := binary.Size(ch)
-	bodySize, err := packedSize(cmd...)
+	body, err := Pack(cmd...)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't compute packed size for message body: %v", err)
+		return nil, fmt.Errorf("couldn't pack message body: %v", err)
 	}
+	bodySize := binary.Size(body)
```

Comment:
isn't this just len(body)?

Reply:
just to be consistent with the `binary.Size(ch) above` I can change it to len if you'd like.

Reply:
Yeah, no need to jump through the type check in binary.Size, use len